content/blue-green-deploys-dbt-snowflake.md

```diff
@@ -0,0 +1,85 @@
+
+Let us consider the following **scenario**.
+
+Present : You have 2 events that are combined into a metric using session_ids. The size of the tables are depicted in yellow notes. 
+
+![scenario-present](https://p91.f3.n0.cdn.getcloudapp.com/items/eDu6Rnop/Image%202020-05-02%20at%209.32.13%20AM.png)
+
+
+Future: Price is being added to `product_event_1`. We need to change the models and have to run a full refresh to calculate the new fields for all rows.
+
+![scenario-future](https://p91.f3.n0.cdn.getcloudapp.com/items/xQuW6OvW/Image%202020-05-02%20at%209.34.04%20AM.png)
+
+The problem is: 
+* Downtime in backfilling/full refreshing production models that take a long time
+*  Toil of managing ad-hoc  jobs for backfilling
+* Test failures are very expensive at this stage
+
+### Its Impact:
+* Reduced downtime
+* Reduced toil
+* Ability to rollback
+
+### Proposed solution:
+
+We can learn from Continuous Delivery best practices that Blue Green deploys are a way to solve this problem. So below is my proposed solution to perform blue/green deployments with dbt and snowflake.
+
+**What are blue-green deploys?**
+Blue-Green deployment approach is one where you have two identical environments, labeled Blue and green respectively. One is *live* for on production, lets call that Green. The other environment is Blue which is used for doing the development/new changes. Once the changes to the Blue deployment are complete and verified then there is a switch over between them. Blue becomes green (active) and green becomes blue (inactive). 
+
+Features from dbt and snowflake which the proposed solution relies on:
+* zero-copy clone
+* materialized models: incremental and insert by period
+
+
+**Infrastructure Setup:**
+1. DBT repo has 2 branches, pre-prod and prod
+2. Green environment for `dbt run` prod branch code with incremental models on a daily/hourly schedule
+3. Blue environment for `dbt run` pre-prod branch code with full refresh models on an external trigger
+4. Inside snowflake `analytics.public` schema is production
+5. Inside snowflake `analytics.public_upcoming` schema is pre-prod
+ 
+Blue DBT runs should only run parts of lineage graph that are heavily incremental and should be selecting models based code diff between pre-prod and prod files.
+
+**The process:**
```

Comment:
I don't think this is that complex, or presented in a complex way, but for some reason it took me two read-throughs to get it. Maybe that's fine or maybe that's just me, but also a diagram for some of the steps for where the data is and what code is being run on each set of data will be useful.

Reply:
I should make it easier for anyone to read. My audience is dbt + snowflake users because I dont think the solution is generally applicable 😿  or is it 🤔